complex/ft_test.c

```diff
@@ -124,6 +125,12 @@ static void ft_cleanup_xcontrol(struct ft_xcontrol *ctrl)
 	memset(ctrl, 0, sizeof *ctrl);
 }
 
+static void ft_cleanup_mr_control(struct ft_mr_control *ctrl)
+{
+	free(ctrl->buf);
```

Comment:
The setup function implied that ctrl->buf could be allocated somewhere else, which makes me suspect that it is also freed somewhere else.  If there is a single pairing of setup/cleanup, then I would remove the checks in setup that determine if buf has already been allocated.